### ISSUE TITLE
async arrow function with space-before-function-paren rule

### ISSUE THREAD

**Issue by jcubic** [eslint/eslint#13983]
**Tell us about your environment**
```
Environment Info:
Node version: v14.3.0
npm version: v6.14.9
Local ESLint version: v7.17.0 (Currently used)
Global ESLint version: Not found
```

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?** No idea, just installed eslint without any extra parser from npm. I'm using JavaScript not TypeScript.

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
  "eslintConfig": {
    "parserOptions": {
      "ecmaVersion": 2020,
      "sourceType": "module",
      "ecmaFeatures": {}
    },
    "env": {
      "browser": true,
      "jest": true,
      "node": true
    },
    "globals": {
      "Promise": true
    },
    "rules": {
      "eqeqeq": "error",
      "curly": "error",
      "no-unreachable": "error",
      "valid-typeof": 0,
      "no-unexpected-multiline": "error",
      "no-regex-spaces": "error",
      "no-irregular-whitespace": "error",
      "no-invalid-regexp": "error",
      "no-inner-declarations": "error",
      "no-func-assign": "error",
      "no-extra-semi": "error",
      "no-extra-boolean-cast": "error",
      "no-debugger": "error",
      "no-dupe-args": "error",
      "no-dupe-keys": "error",
      "no-duplicate-case": "error",
      "no-empty-character-class": "error",
      "no-ex-assign": "error",
      "array-callback-return": "error",
      "no-case-declarations": "error",
      "guard-for-in": "error",
      "no-caller": "error",
      "no-eval": "error",
      "no-extend-native": "error",
      "no-extra-bind": "error",
      "no-fallthrough": "error",
      "no-global-assign": "error",
      "no-implicit-globals": "error",
      "no-implied-eval": "error",
      "no-multi-spaces": "error",
      "no-new-wrappers": "error",
      "no-redeclare": "error",
      "no-self-assign": "error",
      "no-return-assign": "error",
      "no-self-compare": "error",
      "no-throw-literal": "error",
      "no-unused-labels": "error",
      "no-useless-call": "error",
      "no-useless-escape": "error",
      "no-void": "error",
      "no-with": "error",
      "radix": "error",
      "wrap-iife": [
        "error",
        "inside"
      ],
      "yoda": [
        "error",
        "never"
      ],
      "no-catch-shadow": "error",
      "no-delete-var": "error",
      "no-label-var": "error",
      "no-undef-init": "error",
      "no-unused-vars": "error",
      "no-undef": "error",
      "comma-style": [
        "error",
        "last"
      ],
      "comma-dangle": [
        "error",
        "never"
      ],
      "comma-spacing": [
        "error",
        {
          "before": false,
          "after": true
        }
      ],
      "computed-property-spacing": [
        "error",
        "never"
      ],
      "eol-last": [
        "error",
        "always"
      ],
      "func-call-spacing": [
        "error",
        "never"
      ],
      "key-spacing": [
        "error",
        {
          "beforeColon": false,
          "afterColon": true,
          "mode": "strict"
        }
      ],
      "max-len": [
        "error",
        90
      ],
      "max-statements-per-line": "error",
      "new-parens": "error",
      "no-array-constructor": "error",
      "no-lonely-if": "error",
      "no-mixed-spaces-and-tabs": "error",
      "no-multiple-empty-lines": "error",
      "no-new-object": "error",
      "no-tabs": "error",
      "no-trailing-spaces": "error",
      "no-whitespace-before-property": "error",
      "object-curly-spacing": [
        "error",
        "always"
      ],
      "space-before-blocks": "error",
      "keyword-spacing": [
        "error",
        {
          "before": true,
          "after": true
        }
      ],
      "space-in-parens": [
        "error",
        "never"
      ],
      "space-infix-ops": "error",
      "space-before-function-paren": [
        "error",
        "never"
      ],
      "complexity": [
        "error",
        {
          "max": 35
        }
      ],
      "indent": [
        "error",
        4,
        {
          "SwitchCase": 1
        }
      ],
      "linebreak-style": [
        "error",
        "unix"
      ],
      "semi": [
        "error",
        "always"
      ]
    }
  },
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

I'm using ESLit for quote some time in this project but first time used async arrow function:

```
    function InputPort(read) {
        this._read = read;
        this._with_parser = this._with_init_parser.bind(this, async () => {
            if (!this.__parser__ || this.__parser__.__lexer__.peek() === eof) {
                const line = await this._read();
                this.__parser__ = new Parser(line, { env: this });
            }
            return this.__parser__;
        });
    }
```

Found this question on SO [ESLint parsing error when using arrow syntax with async](https://stackoverflow.com/q/44477769/387194). But even adding 2020 as version didn't solved the issue:

```bash
./node_modules/.bin/eslint src/lips.js
```

**What did you expect to happen?**

No error, async arrow function should not have space before parenthesis

ESLint say that valid is:

```
async() => {
```

**What actually happened? Please include the actual, raw output from ESLint.**

```
error  Unexpected space before function parentheses  space-before-function-paren
```

**Are you willing to submit a pull request to fix this bug?**

 I'm not sure if I know how, but if this is a bug in can see if I can look at the code and fix the issue.

**Comment by anikethsaha** [comment]
Thanks for the issue

This is working as expected. 

as you have set `never` option for `space-before-function-paren`,  this would disallows any space before `(`

docs [here](https://eslint.org/docs/rules/space-before-function-paren#never)

**Comment by jcubic** [comment]
Yes but I think that arrow function is exception to to this rules it's global and should affect only `function () {}` here you have function keyword and parenthesis with arrow function it look like bug and the code looks stupid, for function keyword both cases looks reasonable.

**Comment by anikethsaha** [comment]
if you want space between `async` and `(` then you can use `{"anonymous": "never", "named": "never", "asyncArrow": "always"}`  option for this rule instead.

[demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgc3BhY2UtYmVmb3JlLWZ1bmN0aW9uLXBhcmVuOiBbXCJlcnJvclwiLCB7XCJhbm9ueW1vdXNcIjogXCJuZXZlclwiLCBcIm5hbWVkXCI6IFwibmV2ZXJcIiwgXCJhc3luY0Fycm93XCI6IFwiYWx3YXlzXCJ9XSovXG4vKmVzbGludC1lbnYgZXM2Ki9cblxuICAgIGZ1bmN0aW9uIElucHV0UG9ydChyZWFkKSB7XG4gICAgICAgIHRoaXMuX3JlYWQgPSByZWFkO1xuICAgICAgICB0aGlzLl93aXRoX3BhcnNlciA9IHRoaXMuX3dpdGhfaW5pdF9wYXJzZXIuYmluZCh0aGlzLCBhc3luYyAoKSA9PiB7XG4gICAgICAgICAgICBpZiAoIXRoaXMuX19wYXJzZXJfXyB8fCB0aGlzLl9fcGFyc2VyX18uX19sZXhlcl9fLnBlZWsoKSA9PT0gZW9mKSB7XG4gICAgICAgICAgICAgICAgY29uc3QgbGluZSA9IGF3YWl0IHRoaXMuX3JlYWQoKTtcbiAgICAgICAgICAgICAgICB0aGlzLl9fcGFyc2VyX18gPSBuZXcgUGFyc2VyKGxpbmUsIHsgZW52OiB0aGlzIH0pO1xuICAgICAgICAgICAgfVxuICAgICAgICAgICAgcmV0dXJuIHRoaXMuX19wYXJzZXJfXztcbiAgICAgICAgfSk7XG4gICAgfSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Comment by jcubic** [comment]
Works great, will add this issue to StackOverflow as answer.
